pkg/collector/corechecks/snmp/discovery/discovery.go

```diff
@@ -0,0 +1,294 @@
+func (d *Discovery) Start() {
+	log.Debugf("subnet %s: Start discovery", d.config.Network)
+	go d.checkDevices()
+}
+
+// Stop signal discovery to shut down
+func (d *Discovery) Stop() {
+	log.Debugf("subnet %s: Stop discovery", d.config.Network)
+	close(d.stop)
+}
+
+// GetDiscoveredDeviceConfigs returns discovered device configs
+func (d *Discovery) GetDiscoveredDeviceConfigs() []*devicecheck.DeviceCheck {
+	d.discDevMu.RLock()
+	defer d.discDevMu.RUnlock()
+
+	discoveredDevices := make([]*devicecheck.DeviceCheck, 0, len(d.discoveredDevices))
+	for _, device := range d.discoveredDevices {
+		discoveredDevices = append(discoveredDevices, device.deviceCheck)
+	}
+	return discoveredDevices
+}
+
+// Start discovery
+func (d *Discovery) runWorker(w int, jobs <-chan snmpJob) {
+	log.Debugf("subnet %s: Start SNMP worker %d", d.config.Network, w)
+	for {
+		select {
+		case <-d.stop:
+			log.Debugf("subnet %s: Stop SNMP worker %d", d.config.Network, w)
+			return
+		case job := <-jobs:
+			log.Debugf("subnet %s: Handling IP %s", d.config.Network, job.currentIP.String())
+			err := d.checkDevice(job)
+			if err != nil {
+				log.Errorf(err.Error())
+			}
+		}
+	}
+}
+
+func (d *Discovery) checkDevices() {
+	ipAddr, ipNet, err := net.ParseCIDR(d.config.Network)
+	if err != nil {
+		log.Errorf("subnet %s: Couldn't parse SNMP network: %s", d.config.Network, err)
+		return
+	}
+
+	startingIP := ipAddr.Mask(ipNet.Mask)
+
+	configHash := d.config.DiscoveryDigest(d.config.Network)
+	cacheKey := fmt.Sprintf("%s:%s", cacheKeyPrefix, configHash)
+
+	subnet := snmpSubnet{
+		config:         d.config,
+		startingIP:     startingIP,
+		network:        *ipNet,
+		cacheKey:       cacheKey,
+		devices:        map[string]string{},
+		deviceFailures: map[string]int{},
+	}
+
+	d.loadCache(&subnet)
+
+	jobs := make(chan snmpJob)
+	for w := 0; w < d.config.DiscoveryWorkers; w++ {
+		go d.runWorker(w, jobs)
+	}
+
+	discoveryTicker := time.NewTicker(time.Duration(d.config.DiscoveryInterval) * time.Second)
+
+	for {
+		log.Debugf("subnet %s: Run discovery", d.config.Network)
+		startingIP := make(net.IP, len(subnet.startingIP))
+		copy(startingIP, subnet.startingIP)
+		for currentIP := startingIP; subnet.network.Contains(currentIP); incrementIP(currentIP) {
+
+			if ignored := subnet.config.IsIPIgnored(currentIP); ignored {
+				continue
+			}
+
+			jobIP := make(net.IP, len(currentIP))
+			copy(jobIP, currentIP)
+			job := snmpJob{
+				subnet:    &subnet,
+				currentIP: jobIP,
+			}
+			jobs <- job
+
+			select {
+			case <-d.stop:
+				log.Debugf("subnet %s: Stop scheduling devices", d.config.Network)
+				return
+			default:
+			}
+		}
+
+		select {
+		case <-d.stop:
+			log.Debugf("subnet %s: Stop scheduling devices", d.config.Network)
+			return
+		case <-discoveryTicker.C:
+		}
+	}
+}
+
+func (d *Discovery) checkDevice(job snmpJob) error {
+	deviceIP := job.currentIP.String()
+	config := *job.subnet.config // shallow copy
+	config.IPAddress = deviceIP
```

Comment:
Is it safe to change the `IPAddress` of the referenced config? Isn't it shared across all discovery runners?

Reply:
Yes, since we are making a shallow copy at line above. 

Added a test to make sure we are not modifying the initial config: fa287611c

Reply:
Thank you, this made me dive back into shallow / deep copy. As per https://en.wikipedia.org/wiki/Object_copying#Shallow_copy a shallow copy creates a new instance of an object and copy all the values from the original object. If any of the values is a pointer, shallow copy will copy the pointer. 

In `golang` strings are values, so here `config.IPAddress` copies the **value** to `job.subnet.config.IPAddress`. 

If `config` contains any pointer backed type (map, pointer, slice, channel, function and interface) golang copies the pointer, so the copied instance will share the same memory address with the original instance's property. Any change to the copied's property will reflect into the original.

Example in [this playground](https://play.golang.org/p/SSxWYJqIsNW)

Here it is fine as `IPAddress` is a string. We need to ensure that A/ config has no property backed by pointers or B/ we never change properties backed by pointers, considering `config` is copied from all the discovery runners go routines